File: in-class-w5-1011-arithmetic-coding.cpp
#include <bits/stdc++.h>
using namespace std;
struct encodeobj{
	map<char, pair<double, double> > table;
	double value;
};
encodeobj encode(string s) {
	int arr[s.length()];
	for(int q=0; q<s.length(); q++) {
		arr[q] = s[q];
	}
	sort(arr, arr+s.length());
	encodeobj res;
	for(int q=0; q<s.length(); q++) {
		res.table[s[q]] = make_pair(1.0 * (lower_bound(arr, arr+s.length(), s[q])-arr) / s.length(), 1.0 * (upper_bound(arr, arr+s.length(), s[q])-arr) / s.length());
	}
	double low = 0, high = 1;
	for(int q=0; q<s.length(); q++) {
		double range = high - low;
		high = low + range * (upper_bound(arr, arr+s.length(), s[q])-arr) / s.length();
		low = low + range * (lower_bound(arr, arr+s.length(), s[q])-arr) / s.length();
	}
	res.value = low;
	return res;
}
string decode(map<char, pair<double, double> > table, double value) {
	string res = "";
	while(value > 1e-7) {
		for(map<char, pair<double, double> >::iterator it=table.begin(); it!=table.end(); ++it) {
			if(value >= it->second.first && value < it->second.second) {
				res += it->first;
				value -= it->second.first;
				value /= (it->second.second - it->second.first);
				break;
			}
		}
	}
	return res;
}
int main() {
	cout<<fixed<<setprecision(17);
	encodeobj res = encode("BILL GATES");
	cout<<res.value<<endl;
	
	cout<<decode(res.table, res.value)<<endl;
}

File: in-class-w5-1011-arithmetic-coding.js
function encode(s) {
    arr = s.split("").sort();
    table = {};
    for (c of s) {
        table[c] = [arr.indexOf(c) / arr.length, (arr.lastIndexOf(c) + 1) / arr.length];
    }
    low = 0;
    high = 1;
    for (c of s) {
        range = high - low;
        high = low + range * (arr.lastIndexOf(c) + 1) / arr.length;
        low = low + range * arr.indexOf(c) / arr.length;
    }
    return {
        "table": table,
        "value": low
    };
}

function decode(table, value) {
    res = "";
    while (value > 1e-7) {
        for (c in table) {
            if (value >= table[c][0] && value < table[c][1]) {
                res += c;
                value -= table[c][0];
                value /= (table[c][1] - table[c][0]);
                break;
            }
        }
    }
    return res;
}

res = encode("BILL GATES");
console.log("encode table:", res["table"]);
console.log("encode value:", res["value"]);

console.log("decode:", decode(res["table"], res["value"]));
